### PR TITLE
Extension methods for Java classes

### LINKED ISSUE #8821: Extension methods & conversions for Java host types
Based on the https://github.com/enso-org/enso/issues/8346#issuecomment-1898044109 idea.

- #8016

Since #7858 was implemented, we can use `polyglot java import` types (e.g. _host Java classes_) in _type ascriptions_. The _host Java classes_ are recognized in `Meta.type_of` and we can even use them in `case of` statements:
```ruby
polyglot java import java.math.BigDecimal

c x = case x of
  _ : BigDecimal -> "big"
  _ -> "small"
```

However there are still few areas where the support of _host Java classes_ isn't on par with _Enso types_ - one **cannot add extension methods** to them and one **cannot define conversion methods**. Both of these features are technically possible and they would _smooth interoperability_ between Enso and Java when designing the _Enso libraries_.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Addresses one part of #8821 - allows definition and invocation of _extension method_ on Java classes.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides. 
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: test/Base_Tests/src/Data/Polyglot_Spec.enso [hunk 1]
```diff
@@ -42,6 +42,10 @@ add_specs suite_builder = suite_builder.group "Polyglot" group_builder->
         hold.value . should_equal 42
         hold.boxed . should_equal 42
 
+    group_builder.specify "extension method on IntHolder" <|
```
Comment (JaroslavTulach) [inline test/Base_Tests/src/Data/Polyglot_Spec.enso:46]: Hi guys. I welcome suggestions how to expand the test suite to cover everything _extension methods_ related.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java [hunk 2]
```diff
@@ -150,6 +155,14 @@ public static PolyglotCallType getPolyglotCallType(
       UnresolvedSymbol symbol,
       InteropLibrary library,
       MethodResolverNode methodResolverNode) {
+    if (symbol.getScope().getMethodForPolyglot(self, symbol.getName(), false)
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:158]: `getMethodForPolyglot` already returns `Function` type. Thus, the instanceof check here is redundant, and null check should be sufficient.
```suggestion
    if (symbol.getScope().getMethodForPolyglot(self, symbol.getName(), false) != null) { 
```

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java [hunk 3]
```diff
@@ -200,6 +202,44 @@ public void registerPolyglotSymbol(String name, Object sym) {
...
+
+  public void registerPolyglotMethod(
+      Object metaObject, String name, boolean instance, Supplier<Function> fn) {
+    String meta = metaKey(metaObject);
+    var m = polyMethods.get(meta);
+    if (m == null) {
+      m = new HashMap<>();
+      polyMethods.put(meta, m);
+    }
+    m.put(name, new CachingSupplier<>(fn));
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:215]:
```suggestion
    var m = polyMethods.computeIfAbsent(meta, _ -> new HashMap<>());
```

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java [hunk 4]
```diff
@@ -200,6 +202,44 @@ public void registerPolyglotSymbol(String name, Object sym) {
     polyglotSymbols.put(name, sym);
   }
 
+  private final Map<String, Map<String, Supplier<Function>>> polyMethods = new HashMap<>();
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:205]: minor: Although not enforced, we tend to declare fields before methods.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/IrToTruffleUtils.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
+package org.enso.interpreter.runtime;
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/IrToTruffleUtils.java:1]: minor: Put into `org.enso.interpreter.runtime.utils` package.

File: test/Base_Tests/src/Data/Polyglot_Spec.enso (hunk 1 above)
Comment (radeusgd) [reply test/Base_Tests/src/Data/Polyglot_Spec.enso:46]: The 3 tests already there seem to test the fundamental properties. I like that we also ensure to test the static syntax.

Firstly, I would add some tests that check what happens if there is ambiguity - e.g. if one defines an extension method in Enso that 'shadows' a method with the same name that was already on the Java type.

However, this has never worked right for Enso types (the issue #1243 is 3.5 years old and is still persistent - _checked today_). So I'm not sure if it is in the scope of this PR to get this fixed, but in general I think we probably should get this sorted out.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/IrToTruffleUtils.java (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/IrToTruffleUtils.java:1]: I'd like to keep `IrToTruffleUtils` package private and move there some code from `IrToTruffle`. As such both classes need to be in the same package.

File: test/Base_Tests/src/Data/Polyglot_Spec.enso (hunk 1 above)
Comment (radeusgd) [reply test/Base_Tests/src/Data/Polyglot_Spec.enso:46]: These seem less crucial, but we could also add test cases for following edge cases:
- extension method taking no arguments,
- extension method with default arguments,
	- and in particular - with _all_ arguments defaulted.

(all of these could be done both for static and member method if we want to be thorough)

Knowing that there used to be problems with that, I guess it may also be good to check suspended arguments and suspended+defaulted arguments (related bug #8937). However, if I understand correctly there probably should be no reason why these would work differently than for regular Enso methods.

File: test/Base_Tests/src/Data/Polyglot_Spec.enso [hunk 6]
```diff
@@ -42,6 +42,26 @@ add_specs suite_builder = suite_builder.group "Polyglot" group_builder->
...
+        hold = IntHolder.new (6 * 7)
+        v = hold.value_plus 3
+        v . should_equal 45
+
+    group_builder.specify "extension method on IntHolder (static syntax)" <|
+        hold = IntHolder.new (6 * 7)
+        v = IntHolder.value_plus hold 3
```
Comment (JaroslavTulach) [inline test/Base_Tests/src/Data/Polyglot_Spec.enso:53]: With the latest commit af736be149fbb75ec8f57c92c8a1a5a9aa469b72 all tests but this one seem to be passing:
```
    - [FAILED] extension method on IntHolder (static syntax) [98ms]
        Reason: An unexpected panic was thrown: (Type_Error.Error Integer org.enso.base_test_helpers.IntHolder@16120270 '`add`')
        at <enso> Any.value_plus(/enso/test/Base_Tests/src/Data/Polyglot_Spec.enso:78:1-54)
        at <enso> <anonymous><arg-1>(enso/test/Base_Tests/src/Data/Polyglot_Spec.enso:52:13-39)
```
for some reason I have an oversaturated argument there. The static calls were introduced by
- #3764

With 488ca26 the tests seem to work. The code of 488ca26 is however completely horrible and needs to be polished to run on _fast path_ somehow. PS: I still don't understand why #3764 was implemented on the level of `IR` and not on Truffle level and `ModuleScope`s.

File: test/Base_Tests/src/Data/Polyglot_Spec.enso (hunk 1 above)
Comment (radeusgd) [reply test/Base_Tests/src/Data/Polyglot_Spec.enso:46]: Lastly, given the recently implemented #8867, I guess it may be good to also verify if the typecheck works correctly in case of the extensions on Java types.

Of course your static syntax test already covers the 'happy path'.

So I'd 'just' add one more case:
```
Test.expect_panic Type_Error (IntHolder.value_plus 10 12)
```

that checks that we cannot give an Enso `Integer` for the Java `IntHolder` `self`.

Comment (JaroslavTulach) [reply test/Base_Tests/src/Data/Polyglot_Spec.enso:46]: Thanks for the test. Used in e279793

Comment (JaroslavTulach) [reply test/Base_Tests/src/Data/Polyglot_Spec.enso:46]: Yeah, #1243 looks serious - I wonder why it doesn't come up during the weekly planning session. CCing @jdunkerley. I've added `triage` label - maybe it helps.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java:205]: Honestly, the code I have put into `ModuleScope` is so ugly that it deserves a _total rewrite_. With e279793 we seem to have the desirable functionality, but the implementation has to improve significantly.

Pavel, do you think the following would be a fine goal?

**Give every Java class an Enso `Type`**

I am still a bit afraid of doing so, but it would solve a lot of problems. If there was a _Type wrapper around each Java class_ used in Enso, we could easily assign the extension methods the _regular way_ - e.g. without duplicating all these internal fields and register methods. We would just assign them to the appropriate `Type` wrapping the Java class. Moreover Java types would appear in the IDE as types. Something that
- #8009

wanted to achieve a long time ago.

I would just need to _open_ `Type`  _up_ and a properly "cache" the type and java class mapping - guarantee it is 1:1. We probably still need the _eigen type_ concept (even for Java classes) as that's where Enso stores static methods. The current implementation struggles a lot differentiating between static and non-static methods (see current [MethodInvokeNode](https://github.com/enso-org/enso/pull/9129/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R852)). And yes, re-implementing #3764 to register each method only once and _just dispatch properly_ would make me happier, but it is another thing I am hesitating/afraid to do...

**Comment by JaroslavTulach** [comment]
The approach of compiling `Standard` libraries into `runner` executable seems viable - https://github.com/enso-org/enso/pull/9866#issuecomment-2112450583. That however means we don't want to expose Java objects in the system as first class types/instances. If a library compiled by NI sends object to library using regular JVM or vice versa they will not see the other object as being from (own) JVM.

As such I am closing this PR as going in the _wrong direction_.
